adapt for 3.4

bozana opened the pull request:
s. https://github.com/pkp/pkp-lib/issues/8597

bozana commented:
@asmecher, could you please take a look how I implemented the mEDRA plugin migration, s. https://github.com/bozana/medra/blob/3_4_0/MedraPlugin.php#L65 ?
The migration script should be run only once, only first time on `preInstall` of this new generic mEDRA plugin.
I tested it with a local plugin gallery's plugins.xml and there should not be any problems having the new generic plugin mEDRA. Just we need to be careful when we edit the plugins.xml to be aware that there is the old importexport and the new generic mEDRA plugin. For the user this will not be the problem, because the plugins are for different OJS releases and both will/can not be displayed/available for the same OJS installation -- importexport plugin is available for OJS < 3.4 and generic plugin is available for OJS >= 3.4.
I haven't implemented the down() method in the migration script. Do you think I should (because this migration script is run when installing the plugin via plugin gallery)? Or is there maybe a way to warn the user to do the backup before installing the new plugin?

ewhanson commented:
Hey @bozana, I've been reviewing the plugin and had a question about the intended upgrade flow. What is the intended sequence of events for an upgrade? I've been testing it with an upgrade from 3.3 to 3.4. In this case, should it be:

1. Keep old import/export medra plugin
2. Run 3.3 to 3.4 upgrade
3. Remove import/export medra plugin manually (deleting from file system?)
4. Install medra plugin from plugin gallery (or git)
5. At this point the `Installer::preInstall` migration is run

Is that correct? Is there a direct upgrade path using the plugin gallery for the plugin type move (from import/export to generic)?

bozana commented:
@ewhanson,
I thought:
- The user will take the new 3.4 code. This code will not have medra plugin, but the old medra importexport plugin is in the DB.
- Then the user does the upgrade to 3.4.
- Then the user installs the new medra plugin via plugin gallery and in this moment Installer::preInstall migration is run.

I am not sure if Installer::preInstall would be run if the plugin would be installed via git i.e. manually -- I would need to double check. I am not sure if we need to support this, because in README we say that the plugin gallery should be used.
But, maybe according to your experience with publishing services we could/should consider something else too... ?

ewhanson commented on filter/ArticleMedraXmlFilter.php:
```diff
@@ -226,14 +241,10 @@ function createArticleNode($doc, $pubObject) {
-	function createContentItemNode($doc, $issue, $article, $galley, $objectLocalePrecedence) {
+	function createContentItemNode(DOMDocument $doc, Issue $issue, Submission $article, ?Galley $galley, array $objectLocalePrecedence): DOMElement
+	{
+		/** @var PKPNativeImportExportDeployment $deployment */
```
Needs a full name space or use statement.

ewhanson commented on filter/IssueMedraXmlFilter.php:
```diff
@@ -29,40 +40,39 @@ function __construct($filterGroup) {
-		return 'plugins.importexport.medra.filter.IssueMedraXmlFilter';
+	function getRootNodeName(): string
+	{
+		/** @var PKPNativeImportExportDeployment $deployment */
```
Needs full namespace/use statement.

ewhanson commented on MedraExportDeployment.php:
```diff
@@ -0,0 +1,121 @@
+	/**
+	 * Set the import/export plugin.
+	 */
+	function setPlugin(ImportExportPlugin $plugin): void
```
This isn't used anymore as it's assigned directly in the constructor. Could it be removed? If it needs to be kept, `$this->plugin` is of type `MedraExportPlugin` in the constructor.

bozana replied:
OK, yes, I will remove this and setContext() function...

ewhanson commented on MedraExportPlugin.php:
```diff
@@ -0,0 +1,361 @@
+    private function _updateDepositStatus(DataObject $object, string $status, string $failedMsg = null)
+    {
+        if ($object instanceof Submission) {
+			/** @var Publication $object */
```
`Publication` needs a full namespace or use statement.

ewhanson commented on MedraPlugin.php:
```diff
@@ -0,0 +1,394 @@
+        $schema = &$args[0];
+
+        $settings = [
+            $this->getFailedMsgSettingName(),
```
Are there any failed messages that would need to be carried forward as part of the migration? I didn't see anything about them there.

bozana replied:
Yes, there have been none till now -- we would only display the notification, but now, that we cannot easily display notifications, we save them.

ewhanson commented on MedraPlugin.php:
```diff
@@ -0,0 +1,394 @@
+            $pluginPathName = 'MedraExportPlugin';
+            $this->_exportPlugin = PluginRegistry::getPlugin($pluginCategory, $pluginPathName);
+            // If being run from CLI, there is no context, so plugin initialization would not have been fired
+            // Note: from the release 3.4 there is no possibility to run the plugin from CLI
```
I think this might also be the case if things are invoked from a job, but I'm not 100%.

ewhanson commented on classes/MedraWebservice.php:
```diff
@@ -150,8 +142,8 @@ function _doRequest($action, $arg, $attachment = null) {
 
-		if ($result === true && ($status = $response->getStatusCode()) != MEDRA_WS_RESPONSE_OK) {
-			$result = 'OJS-mEDRA: Expected ' . MEDRA_WS_RESPONSE_OK . ' response code, got ' . $status . ' instead.';
+		if ($result === true && ($status = $response->getStatusCode()) != self::MEDRA_WS_RESPONSE_OK) {
```
I think `$result` here will always be true because it's only set to something other than `true` inside the catch block.

ewhanson commented on filter/O4DOIXmlFilter.php:
```diff
@@ -1,125 +1,139 @@
-	function createRootNode($doc, $rootNodeName) {
+	function createRootNode(DOMDocument $doc, string $rootNodeName): DOMElement
+	{
+		/** @var PKPNativeImportExportDeployment $deployment */
```
Needs full namespace/use statement.

ewhanson commented on filter/O4DOIXmlFilter.php:
```diff
@@ -201,20 +213,17 @@ function createSerialWorkNode($doc, $journalLocalePrecedence) {
+	 * @param string $titleType O4DOI_TITLE_TYPE_*
 	 */
-	function createTitleNode($doc, $locale, $localizedTitle, $titleType, $localizedSubtitle = null) {
+	function createTitleNode( DOMDocument$doc, string $locale, string $localizedTitle, string $titleType, string $localizedSubtitle = null): DOMElement
```
There's an additional space before `DOMDocument` and a missing space between that and `$doc`. Doesn't appear to be causing an issue, though.

ewhanson commented:
Hey @bozana, I've reviewed the plugin and it all looks good! I just left a few comments in the code and had a few other general notes.

> I am not sure if Installer::preInstall would be run if the plugin would be installed via git i.e. manually -- I would need to double check. I am not sure if we need to support this, because in README we say that the plugin gallery should be used

For this, `Installer::preInstall` is not run unless it's installed via the plugin gallery or `tar.gz` upload. This means it won't be possible to directly run the migration if it's installed via git. Not sure if this would ever come up outside of development, though. Also will there be any issue in the future if another migration is added? Should there be any plugin version related information about which versions of the plugin this migration should be run for?

I didn't see any references to `crDepositErrors.tpl` anywhere in the code. Is this still used?

There's also a mix of tabs and spaces throughout the code.

Thanks! Let me know if you have any questions.

bozana commented:
Hi @ewhanson, thanks a lot!
I will double check with @asmecher, if I should consider installing the plugin via git, but somehow I do not think so -- it should be installed via plugin gallery.
This migration script should be run only the first time this new generic plugin is installed. It does not matter which version the plugin is in -- the user could update from X to 3.4 or 3.5 or Y so that we do not know in which version of this new generic plugin they will start using it. But once they start using it, this migration script should not be run again when they upgrade the plugin.
If another migration is added, it will need to consider the right situation (e.g. for which version etc.) then.

You are right -- I thought I removed that file crDepositErrors.tpl. Thanks! :-)

And yes, you are right -- I forgot that earlier the tabs were used -- I forgot that to replace them. I now replaced the indentation to spaces. Thanks! :-)

And thanks for all other comments and tests!
I will consider all your comments and will then make the release...

bozana commented:
Hi @ewhanson, I realized that there is no harm in allowing users to execute the migration script from CLI -- it could even help if something goes wrong when installing from the plugin gallery. Thus, now I check if plugin is installed in order to run the script in the function preInstall() (and not in the function up() as earlier).
I will try to release the plugin today. Fingers crossed that everything works as it should... :-)
